transport/p2p/src/lib.rs

```diff
@@ -28,35 +29,67 @@ pub mod swarm;
 /// P2P behavior definitions for the transport level interactions not related to the HOPR protocol
 mod behavior;
 
-use futures::Stream;
+use futures::{AsyncRead, AsyncWrite, Stream};
 use libp2p::{swarm::NetworkBehaviour, StreamProtocol};
 use serde::{Deserialize, Serialize};
 use std::fmt::Debug;
 
 use hopr_internal_types::legacy;
-use hopr_internal_types::protocol::Acknowledgement;
 use hopr_transport_identity::PeerId;
 use hopr_transport_network::messaging::ControlMessage;
 use hopr_transport_network::network::NetworkTriggeredEvent;
 use hopr_transport_network::ping::PingQueryReplier;
 use hopr_transport_protocol::PeerDiscovery;
 
-use crate::constants::{
-    HOPR_ACKNOWLEDGE_PROTOCOL_V_0_1_0, HOPR_HEARTBEAT_PROTOCOL_V_0_1_0, HOPR_MESSAGE_PROTOCOL_V_0_1_0,
-    HOPR_TICKET_AGGREGATION_PROTOCOL_V_0_1_0,
-};
+use crate::constants::{HOPR_HEARTBEAT_PROTOCOL_V_0_1_0, HOPR_TICKET_AGGREGATION_PROTOCOL_V_0_1_0};
 
 pub const MSG_ACK_TIMEOUT: std::time::Duration = std::time::Duration::from_secs(10);
 
 /// `Ping` protocol base type for the ping operation
 #[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
 pub struct Ping(pub ControlMessage);
 
-/// `Ping` protocol base type for the pong operation
-
+/// `Pong` protocol base type for the pong operation
 #[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
 pub struct Pong(pub ControlMessage, pub String);
 
+// Control object for the streams over the HOPR protocols
+#[derive(Clone)]
+pub struct HoprStreamProtocolControl {
+    control: libp2p_stream::Control,
+    protocol: StreamProtocol,
+}
+
+impl HoprStreamProtocolControl {
+    pub fn new(control: libp2p_stream::Control, protocol: &'static str) -> Self {
+        Self {
+            control,
+            protocol: StreamProtocol::new(protocol),
+        }
+    }
+}
+
+impl std::fmt::Debug for HoprStreamProtocolControl {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("HoprStreamProtocolControl")
+            .field("protocol", &self.protocol)
+            .finish()
+    }
+}
+
+#[async_trait::async_trait]
+impl hopr_transport_protocol::stream::BidirectionalStreamControl for HoprStreamProtocolControl {
+    fn accept(
+        mut self,
+    ) -> Result<impl Stream<Item = (PeerId, impl AsyncRead + AsyncWrite + Send)> + Send, impl std::error::Error> {
+        self.control.accept(self.protocol)
+    }
+
+    async fn open(mut self, peer: PeerId) -> Result<impl AsyncRead + AsyncWrite + Send, impl std::error::Error> {
+        self.control.open_stream(peer, self.protocol).await
+    }
```

Comment:
Consider later moving the `AsyncRead + AsyncWrite + Send` into a template parameter on the type, the same for the Error.

Reply:
I will keep these unresolved to come back to them in a separate PR.